src/lib/utils/thorchain/memo/addFinalAssetLimitToMemo.ts

```diff
@@ -0,0 +1,70 @@
+import { assertAndProcessMemo } from './assertAndProcessMemo'
+import { MEMO_PART_DELIMITER } from './constants'
+
+type AddLimitToMemoArgs = {
+  /**
+   * A THORChain memo, with or without a limit/MinOut.
+   */
+  memo: string
+  /**
+   * The final asset amount out to be added to the memo. THOR Aggregator will use the last two numbers as exponents.
+   */
+  finalAssetAmountOut: string
+}
+
+/**
+ * A simple util that adds a final asset limit to a THORChain memo, if the memo doesn't already have one.
+ */
+export const addFinalAssetLimitToMemo = ({ memo, finalAssetAmountOut }: AddLimitToMemoArgs) => {
+  const memoParts = memo.split(MEMO_PART_DELIMITER)
+  const [action] = memoParts
+
+  if (!action) throw new Error(`action is required in memo: ${memo}`)
+
+  let updatedMemo = memo
+
+  switch (action.toLowerCase()) {
+    case 'swap':
+    case '=':
+    case 's': {
+      // SWAP:ASSET:DESTADDR:LIM/INTERVAL/QUANTITY:AFFILIATE:FEE:AGGREGATOR:FINALASSETCONTRACT:FINALASSETAMOUNTOUT
+      const [
+        ,
+        asset,
+        destAddr,
+        limit,
+        affiliate,
+        fee,
+        aggregator,
+        finalAssetContract,
+        _finalAssetAmountOut,
+      ] = memoParts
+      if (!_finalAssetAmountOut) {
```

Comment:
suggestion (preferably blocking): we should check for `aggregator` and `finalAssetContract` as discriminators before running this bit, since these should already be present at this stage.

- If this is ever called on a regular (not DEX) memo, then we would end up with something like `SWAP:ASSET:DESTADDR:LIM/INTERVAL/QUANTITY:AFFILIATE:FEE:::FINALASSETAMOUNTOUT`.
- If this is a `SwapOut` memo but aggregator/final asset contract address parts are missing, then this is definitely a bug

Reply:
Added some throws in [2a6fc6a](https://github.com/shapeshift/web/pull/7146/commits/2a6fc6a57521f17fb2fa9ee67d3d46f222c6e53c)